env/posix/ocf_env.h

```diff
@@ -49,31 +55,72 @@ typedef uint64_t u64;
 typedef uint64_t sector_t;
 
 #define __packed	__attribute__((packed))
-#define __aligned(x)	__attribute__((aligned(x)))
 
 #define likely(cond)       __builtin_expect(!!(cond), 1)
 #define unlikely(cond)     __builtin_expect(!!(cond), 0)
 
-#define min(a,b) MIN(a,b)
-
-#define OCF_ALLOCATOR_NAME_MAX 128
-
-#define PAGE_SIZE 4096
-
-/* *** DEBUGING *** */
+/* MEMORY MANAGEMENT */
+#define ENV_MEM_NORMAL	0
+#define ENV_MEM_NOIO	0
+#define ENV_MEM_ATOMIC	0
 
+/* DEBUGING */
 #define ENV_WARN(cond, fmt...)		printf(fmt)
 #define ENV_WARN_ON(cond)		;
 #define ENV_WARN_ONCE(cond, fmt...)	ENV_WARN(cond, fmt)
 
 #define ENV_BUG()			assert(0)
 #define ENV_BUG_ON(cond)		do { if (cond) ENV_BUG(); } while (0)
 
-/* *** MEMORY MANAGEMENT *** */
-#define ENV_MEM_NORMAL	0
-#define ENV_MEM_NOIO	0
-#define ENV_MEM_ATOMIC	0
+/* MISC UTILITIES */
+#define container_of(ptr, type, member) ({          \
+	const typeof(((type *)0)->member)*__mptr = (ptr);    \
+	(type *)((char *)__mptr - offsetof(type, member)); })
+
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof(*(x)))
+
+/* STRING OPERATIONS */
+#ifndef UNIT_TEST
+#define env_memcpy(dest, dmax, src, slen) ({ \
+		memcpy(dest, src, min(dmax, slen)); \
+		0; \
+	})
+#endif
+#define env_memset(dest, dmax, val) ({ \
+		memset(dest, val, dmax); \
+		0; \
+	})
+#define env_memcmp(s1, s1max, s2, s2max, diff) ({ \
+		*diff = memcmp(s1, s2, min(s1max, s2max)); \
+		0; \
+	})
+#define env_strdup strndup
+#define env_strnlen(s, smax) strnlen(s, smax)
+#define env_strncmp strncmp
+#define env_strncpy(dest, dmax, src, slen) ({ \
+		strncpy(dest, src, min(dmax - 1, slen)); \
+		dest[dmax - 1] = '\0'; \
+		0; \
+	})
+
+#ifdef UNIT_TEST
+int env_memcpy(void *dest, size_t destsz, const void * src, size_t count)
```

Comment:
isn't this functionally identical to the other version (line 84) ?

Reply:
Indeed, but UT can't handle this function as a macro

Reply:
fixed